server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1285,19 +1286,16 @@ private BlobStoreIndexShardSnapshot loadShardSnapshot(BlobContainer shardContain
      * @param blobs list of blobs in repository
      * @return tuple of BlobStoreIndexShardSnapshots and the last snapshot index generation
      */
-    private Tuple<BlobStoreIndexShardSnapshots, Long> buildBlobStoreIndexShardSnapshots(Set<String> blobs, BlobContainer shardContainer) {
+    private Tuple<BlobStoreIndexShardSnapshots, Long> buildBlobStoreIndexShardSnapshots(Set<String> blobs, BlobContainer shardContainer)
+            throws IOException {
         long latest = latestGeneration(blobs);
         if (latest >= 0) {
-            try {
-                final BlobStoreIndexShardSnapshots shardSnapshots =
-                    indexShardSnapshotsFormat.read(shardContainer, Long.toString(latest));
-                return new Tuple<>(shardSnapshots, latest);
-            } catch (IOException e) {
-                final String file = SNAPSHOT_INDEX_PREFIX + latest;
-                logger.warn(() -> new ParameterizedMessage("failed to read index file [{}]", file), e);
-            }
-        } else if (blobs.isEmpty() == false) {
-            logger.warn("Could not find a readable index-N file in a non-empty shard snapshot directory [{}]", shardContainer.path());
+            final BlobStoreIndexShardSnapshots shardSnapshots = indexShardSnapshotsFormat.read(shardContainer, Long.toString(latest));
```

Comment:
Should we also beef this up in the case of the `index-N` file is just missing?

Because looking at `buildBlobStoreIndexShardSnapshots()` callers, `snapshotShard()` passes a list of blobs which names start with `index-` so it will never trigger the `anyMath()` condition in case of missing `index-N`. If two snapshots are created for a shard and if the `index-N` is somewhat deleted after that, the next snapshot creation will silently create a new `index-0` with previous snapshot data missing (but previous snapshots will still appear as SUCCESS).

Reply:
@tlrx you're right that this only fully works in case of deletes not snapshot creations.

For the create side of things we'd have to either do a more expensive list and list out all contents of the shard folder or pass the `RepositoryData` (or something holding similar information) to check whether there should already be an `index-N` in the path I guess?

I'd just do that in #46250 anyway (via to uuid suffix index- blobs, there's hard enforcement on existence) but would leave the solution here as a best-effort solution maybe. Not sure it's worth it adding more complexity to this code if it goes away soon anyway?

Reply:
Can you please update the PR title+desc then to not mention snapshot creation?

Concerning deletions, this PR stops deleting blobs but we don't really fail the snapshot deletion neither we report a failure to the user, right?

Reply:
@tlrx we still ensure safe handling of the corrupted meta case in case of snapshot create. The title doesn't claim anything beyond that :D  But I adjusted the wording in the description :)

> Concerning deletions, this PR stops deleting blobs but we don't really fail the snapshot deletion neither we report a failure to the user, right?

Yea we only log and don't delete physical files for the blob but don't respond with an error. I'm wondering if we should change that ... the problem is that it's been our API behavior forever to not return an error to the user here.
And it sort of makes sense because we already deleted the snapshot from the root level metadata at this point so the users couldn't retry that delete anyway (would just return a missing snapshot error on the retry).

Reply:
> I'd just do that in #46250 anyway (via to uuid suffix index- blobs, there's hard enforcement on existence) but would leave the solution here as a best-effort solution maybe. Not sure it's worth it adding more complexity to this code if it goes away soon anyway?

Ok, we should think of adding tests for this in #46250 if there are none :)

Thanks for the detailed explanations. I'm going to take this PR as a best effort until #46250 takes care of missing shard level index-N files during snapshot creation and until we have shard level blobs clean ups.

Reply:
Thanks!

> Ok, we should think of adding tests for this in #46250 if there are none :)

You caught me :D Will do.